File: kirikiri-tjs/Go/tjs_base_parser.go
package parser

import (
	"strings"

	"github.com/antlr/antlr4/runtime/Go/antlr"
)

// TJSBaseParser implementation.
type TJSBaseParser struct {
	*antlr.BaseParser
}

func (p *TJSBaseParser) notOpenBraceAndNotFunction() bool {
	nextTokenType := p.GetTokenStream().LT(1).GetTokenType()
	return nextTokenType != TJSParserOpenBrace && nextTokenType != TJSParserFunction
}

func (p *TJSBaseParser) closeBrace() bool {
	return p.GetTokenStream().LT(1).GetTokenType() == TJSParserCloseBrace
}

// Returns true if on the current index of the parser's
// token stream a token exists on the Hidden channel which
// either is a line terminator, or is a multi line comment that
// contains a line terminator.
func (p *TJSBaseParser) lineTerminatorAhead() bool {
	// Get the token ahead of the current index.
	possibleIndexEosToken := p.GetCurrentToken().GetTokenIndex() - 1
	ahead := p.GetTokenStream().Get(possibleIndexEosToken)

	if ahead.GetChannel() != antlr.LexerHidden {
		// We're only interested in tokens on the HIDDEN channel.
		return true
	}

	if ahead.GetTokenType() == TJSParserLineTerminator {
		// There is definitely a line terminator ahead.
		return true
	}

	if ahead.GetTokenType() == TJSParserWhiteSpaces {
		// Get the token ahead of the current whitespaces.
		possibleIndexEosToken = p.GetCurrentToken().GetTokenIndex() - 2
		ahead = p.GetTokenStream().Get(possibleIndexEosToken)
	}

	// Get the token's text and type.
	text := ahead.GetText()
	_type := ahead.GetTokenType()

	// Check if the token is, or contains a line terminator.
	return (_type == TJSParserMultiLineComment && (strings.Contains(text, "\r") || strings.Contains(text, "\n"))) ||
		(_type == TJSParserLineTerminator)
}

File: kirikiri-tjs/TypeScript/TJSBaseParser.ts
import { CommonTokenStream, Parser, Lexer } from "antlr4ts";
//import * as TJSParser from "./TJSParser";
const TJSParser = require("./TJSParser");
export abstract class TJSBaseParser extends Parser {
    constructor(input: CommonTokenStream) {
        super(input)
    }
    notOpenBraceAndNotFunction() {
        const nextTokenType = this._input.LT(1).type;
        return (
            nextTokenType !== TJSParser.OpenBrace &&
            nextTokenType !== TJSParser.Function
        );
    }

    closeBrace() {
        return this._input.LT(1).type === TJSParser.CloseBrace;
    }
    lineTerminatorAhead() {
        let possibleIndexEosToken = this.getCurrentToken().tokenIndex - 1;
        let ahead = this._input.get(possibleIndexEosToken);
        if (ahead.channel !== Lexer.HIDDEN) {
            return false;
        }

        if (ahead.type === TJSParser.LineTerminator) {
            return true;
        }

        if (ahead.type === TJSParser.WhiteSpaces) {
            possibleIndexEosToken = this.getCurrentToken().getTokenIndex() - 2;
            ahead = this._input.get(possibleIndexEosToken);
        }

        const text = ahead.type;
        const type = ahead.type;

        return (
            (type === TJSParser.MultiLineComment &&
                (text.includes("\r") || text.includes("\n"))) ||
            type === TJSParser.LineTerminator
        );
    }
}
